File: backend/src/services/codeService.js
/**
 * Code Intelligence Service
 * Provides efficient loading, indexing, and querying of medical codes (CPT, HCPCS, ICD-10/Dx, PCS)
 * Supports loading from chunked JSON files for better memory management
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { Code } from '../models/Code.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

class CodeService {
  constructor() {
    this.codes = [];
    this.codeIndex = new Map();      // code -> full object
    this.typeIndex = new Map();       // type -> [codes]
    this.searchIndex = [];            // Array for text search
    this._isLoaded = false;
    this.loadError = null;
    this.manifest = null;            // Manifest info when loading chunks
  }

  /**
   * Load and index codes - supports both chunked and single file loading
   */
  async loadCodes() {
    if (this._isLoaded) return;

    try {
      const chunksDir = path.join(__dirname, '..', '..', 'data', 'codes_chunks');
      const manifestPath = path.join(chunksDir, 'manifest.json');
      
      // Check if chunked files exist
      if (fs.existsSync(manifestPath)) {
        await this.loadFromChunks(chunksDir, manifestPath);
      } else {
        // Fallback to single file loading
        await this.loadFromSingleFile();
      }
      
      this._isLoaded = true;
    } catch (error) {
      console.error('Error loading codes:', error.message);
      this.loadError = error;
      throw error;
    }
  }

  /**
   * Load codes from chunked JSON files
   */
  async loadFromChunks(chunksDir, manifestPath) {
    console.log('Loading codes from chunked files...');
    const startTime = Date.now();
    
    // Read manifest
    const manifestData = fs.readFileSync(manifestPath, 'utf8');
    this.manifest = JSON.parse(manifestData);
    
    console.log(`Manifest: ${this.manifest.chunkCount} chunks, ${this.manifest.totalCodes} total codes`);
    
    // Load each chunk
    let loadedCodes = 0;
    for (let i = 0; i < this.manifest.chunks.length; i++) {
      const chunk = this.manifest.chunks[i];
      const chunkPath = path.join(chunksDir, chunk.fileName);
      
      const chunkStart = Date.now();
      const chunkData = fs.readFileSync(chunkPath, 'utf8');
      const chunkCodes = JSON.parse(chunkData);
      
      // Add to main codes array
      this.codes.push(...chunkCodes);
      loadedCodes += chunkCodes.length;
      
      console.log(`  Loaded ${chunk.fileName}: ${chunkCodes.length} codes (${Date.now() - chunkStart}ms)`);
    }
    
    console.log(`Loaded ${loadedCodes} codes from ${this.manifest.chunkCount} chunks in ${Date.now() - startTime}ms`);
    
    // Build indexes
    this.buildIndexes();
    
    console.log(`Total loading + indexing: ${Date.now() - startTime}ms`);
  }

  /**
   * Load codes from single JSON file (legacy/fallback)
   */
  async loadFromSingleFile() {
    const dataPath = path.join(__dirname, '..', '..', 'data', 'codes_2025.json');
    console.log('Loading codes from single file:', dataPath);
    
    const startTime = Date.now();
    
    // Read and parse JSON
    const rawData = fs.readFileSync(dataPath, 'utf8');
    this.codes = JSON.parse(rawData);
    
    console.log(`Loaded ${this.codes.length} codes in ${Date.now() - startTime}ms`);
    
    // Build indexes
    this.buildIndexes();
    
    console.log(`Indexing complete in ${Date.now() - startTime}ms`);
  }

  /**
   * Build in-memory indexes for fast lookup
   */
  buildIndexes() {
    console.log('Building indexes...');
    
    // Clear existing indexes
    this.codeIndex.clear();
    this.typeIndex.clear();
    this.searchIndex = [];
    
    for (const codeObj of this.codes) {
      // Index by code
      this.codeIndex.set(codeObj.code, codeObj);
      
      // Index by type
      const type = this.normalizeType(codeObj.type);
      if (!this.typeIndex.has(type)) {
        this.typeIndex.set(type, []);
      }
      this.typeIndex.get(type).push(codeObj);
      
      // Build search index (code + description tokens)
      this.searchIndex.push({
        code: codeObj.code,
        searchText: `${codeObj.code} ${codeObj.description}`.toLowerCase(),
        type: type,
      });
    }
    
    console.log('Index stats:');
    console.log('  - Code index size:', this.codeIndex.size);
    console.log('  - Types:', Array.from(this.typeIndex.keys()));
    for (const [type, codes] of this.typeIndex) {
      console.log(`    - ${type}: ${codes.length} codes`);
    }
  }

  /**
   * Normalize type name for consistent querying
   */
  normalizeType(type) {
    if (!type) return 'OTHER';
    const upperType = type.toUpperCase();
    if (upperType === 'DX') return 'ICD10';
    if (upperType === 'PCS') return 'ICD10-PCS';
    return upperType;
  }

  /**
   * Get all codes with pagination and filtering
   */
  getAllCodes(options = {}) {
    const { 
      limit = 50, 
      offset = 0, 
      type = null,
      sortBy = 'code',
      sortOrder = 'asc'
    } = options;

    let results = type 
      ? (this.typeIndex.get(type.toUpperCase()) || [])
      : this.codes;

    // Sort
    results = [...results].sort((a, b) => {
      const aVal = a[sortBy] || '';
      const bVal = b[sortBy] || '';
      const comparison = aVal.toString().localeCompare(bVal.toString());
      return sortOrder === 'desc' ? -comparison : comparison;
    });

    // Paginate
    const paginated = results.slice(offset, offset + limit);

    return {
      codes: paginated.map(c => this.formatCodeSummary(c)),
      total: results.length,
      limit,
      offset,
      hasMore: offset + limit < results.length,
    };
  }

  /**
   * Get a single code by code string
   */
  getCode(code) {
    const codeObj = this.codeIndex.get(code) || this.codeIndex.get(code.toUpperCase());
    if (!codeObj) return null;
    return this.formatCodeDetail(codeObj);
  }

  /**
   * Search codes by query string
   */
  searchCodes(query, options = {}) {
    const { limit = 50, type = null } = options;
    
    if (!query || query.trim().length < 2) {
      return { codes: [], total: 0, query };
    }

    const searchTerm = query.toLowerCase().trim();
    const terms = searchTerm.split(/\s+/);
    
    // Score-based search
    const results = [];
    
    for (const item of this.searchIndex) {
      // Filter by type if specified
      if (type && item.type !== type.toUpperCase()) continue;
      
      let score = 0;
      
      // Exact code match gets highest score
      if (item.code.toLowerCase() === searchTerm) {
        score = 100;
      } else if (item.code.toLowerCase().includes(searchTerm)) {
        score = 80;
      }
      
      // Term matching in description
      for (const term of terms) {
        if (item.searchText.includes(term)) {
          score += 10;
        }
      }
      
      if (score > 0) {
        results.push({ code: item.code, score });
      }
    }
    
    // Sort by score descending
    results.sort((a, b) => b.score - a.score);
    
    // Get full code objects for top results
    const topResults = results.slice(0, limit).map(r => {
      const codeObj = this.codeIndex.get(r.code);
      return this.formatCodeSummary(codeObj);
    });

    return {
      codes: topResults,
      total: results.length,
      query,
    };
  }

  /**
   * Format code for summary listing
   */
  formatCodeSummary(codeObj) {
    try {
      const code = Code.fromRaw(codeObj);
      return code.toSummary();
    } catch (error) {
      // Fallback formatting
      return {
        code: codeObj.code,
        description: codeObj.description,
        category: codeObj.labels?.[0] || this.normalizeType(codeObj.type),
        type: this.normalizeType(codeObj.type),
        labels: codeObj.labels || [],
      };
    }
  }

  /**
   * Format code for detailed view with payments
   */
  formatCodeDetail(codeObj) {
    try {
      const code = Code.fromRaw(codeObj);
      return code.toDetail();
    } catch (error) {
      // Fallback formatting
      return {
        code: codeObj.code,
        description: codeObj.description,
        category: codeObj.labels?.[0] || this.normalizeType(codeObj.type),
        type: this.normalizeType(codeObj.type),
        labels: codeObj.labels || [],
        payments: { IPPS: 0, HOPD: 0, ASC: 0, OBL: 0 },
        optional: {},
        rawMetadata: codeObj.metadata,
      };
    }
  }

  /**
   * Get statistics about loaded codes
   */
  getStats() {
    const stats = {
      totalCodes: this.codes.length,
      isLoaded: this._isLoaded,
      types: {},
      loadMethod: this.manifest ? 'chunked' : 'single-file',
    };
    
    if (this.manifest) {
      stats.chunks = {
        count: this.manifest.chunkCount,
        targetSizeMB: this.manifest.targetChunkSizeMB,
        createdAt: this.manifest.createdAt,
      };
    }
    
    for (const [type, codes] of this.typeIndex) {
      stats.types[type] = codes.length;
    }
    
    return stats;
  }

  /**
   * Check if service is ready
   */
  isReady() {
    return this._isLoaded && !this.loadError;
  }
}

// Singleton instance
const codeService = new CodeService();

export default codeService;
export { CodeService };

File: backend_python/app/services/code_service.py
"""
Code Intelligence Service
Provides efficient loading, indexing, and querying of medical codes (CPT, HCPCS, ICD-10/Dx, PCS)
Supports loading from chunked JSON files for better memory management
"""

import json
import os
from typing import Dict, List, Optional, Any
from pathlib import Path

from app.models.code import Code


class CodeService:
    """Code service for managing medical codes"""
    
    def __init__(self):
        self.codes: List[Dict[str, Any]] = []
        self.code_index: Dict[str, Dict[str, Any]] = {}  # code -> full object
        self.type_index: Dict[str, List[Dict[str, Any]]] = {}  # type -> [codes]
        self.search_index: List[Dict[str, Any]] = []  # Array for text search
        self._is_loaded = False
        self.load_error: Optional[Exception] = None
        self.manifest: Optional[Dict[str, Any]] = None  # Manifest info when loading chunks
    
    def _get_data_path(self) -> Path:
        """Get the path to the data directory"""
        # Look for data relative to the backend_python directory
        current_dir = Path(__file__).parent.parent.parent
        data_dir = current_dir / "data"
        
        # If not found, check the original backend's data directory
        if not data_dir.exists():
            data_dir = current_dir.parent / "backend" / "data"
        
        return data_dir
    
    async def load_codes(self) -> None:
        """Load and index codes - supports both chunked and single file loading"""
        if self._is_loaded:
            return
        
        try:
            data_path = self._get_data_path()
            chunks_dir = data_path / "codes_chunks"
            manifest_path = chunks_dir / "manifest.json"
            
            # Check if chunked files exist
            if manifest_path.exists():
                await self._load_from_chunks(chunks_dir, manifest_path)
            else:
                # Fallback to single file loading
                await self._load_from_single_file(data_path)
            
            self._is_loaded = True
        except Exception as error:
            print(f"Error loading codes: {error}")
            self.load_error = error
            raise
    
    async def _load_from_chunks(self, chunks_dir: Path, manifest_path: Path) -> None:
        """Load codes from chunked JSON files"""
        import time
        print("Loading codes from chunked files...")
        start_time = time.time()
        
        # Read manifest
        with open(manifest_path, "r", encoding="utf-8") as f:
            self.manifest = json.load(f)
        
        print(f"Manifest: {self.manifest['chunkCount']} chunks, {self.manifest['totalCodes']} total codes")
        
        # Load each chunk
        loaded_codes = 0
        for chunk in self.manifest["chunks"]:
            chunk_path = chunks_dir / chunk["fileName"]
            chunk_start = time.time()
            
            with open(chunk_path, "r", encoding="utf-8") as f:
                chunk_codes = json.load(f)
            
            # Add to main codes array
            self.codes.extend(chunk_codes)
            loaded_codes += len(chunk_codes)
            
            print(f"  Loaded {chunk['fileName']}: {len(chunk_codes)} codes ({int((time.time() - chunk_start) * 1000)}ms)")
        
        print(f"Loaded {loaded_codes} codes from {self.manifest['chunkCount']} chunks in {int((time.time() - start_time) * 1000)}ms")
        
        # Build indexes
        self._build_indexes()
        
        print(f"Total loading + indexing: {int((time.time() - start_time) * 1000)}ms")
    
    async def _load_from_single_file(self, data_path: Path) -> None:
        """Load codes from single JSON file (legacy/fallback)"""
        import time
        codes_file = data_path / "codes_2025.json"
        print(f"Loading codes from single file: {codes_file}")
        
        start_time = time.time()
        
        # Read and parse JSON
        with open(codes_file, "r", encoding="utf-8") as f:
            self.codes = json.load(f)
        
        print(f"Loaded {len(self.codes)} codes in {int((time.time() - start_time) * 1000)}ms")
        
        # Build indexes
        self._build_indexes()
        
        print(f"Indexing complete in {int((time.time() - start_time) * 1000)}ms")
    
    def _build_indexes(self) -> None:
        """Build in-memory indexes for fast lookup"""
        print("Building indexes...")
        
        # Clear existing indexes
        self.code_index.clear()
        self.type_index.clear()
        self.search_index = []
        
        for code_obj in self.codes:
            # Index by code
            self.code_index[code_obj["code"]] = code_obj
            
            # Index by type
            code_type = self._normalize_type(code_obj.get("type"))
            if code_type not in self.type_index:
                self.type_index[code_type] = []
            self.type_index[code_type].append(code_obj)
            
            # Build search index (code + description tokens)
            self.search_index.append({
                "code": code_obj["code"],
                "search_text": f"{code_obj['code']} {code_obj.get('description', '')}".lower(),
                "type": code_type,
            })
        
        print("Index stats:")
        print(f"  - Code index size: {len(self.code_index)}")
        print(f"  - Types: {list(self.type_index.keys())}")
        for type_name, codes in self.type_index.items():
            print(f"    - {type_name}: {len(codes)} codes")
    
    def _normalize_type(self, code_type: Optional[str]) -> str:
        """Normalize type name for consistent querying"""
        if not code_type:
            return "OTHER"
        upper_type = code_type.upper()
        if upper_type == "DX":
            return "ICD10"
        if upper_type == "PCS":
            return "ICD10-PCS"
        return upper_type
    
    def get_all_codes(
        self,
        limit: int = 50,
        offset: int = 0,
        code_type: Optional[str] = None,
        sort_by: str = "code",
        sort_order: str = "asc"
    ) -> Dict[str, Any]:
        """Get all codes with pagination and filtering"""
        if code_type:
            results = self.type_index.get(code_type.upper(), []).copy()
        else:
            results = self.codes.copy()
        
        # Sort
        reverse = sort_order == "desc"
        results.sort(key=lambda x: str(x.get(sort_by, "")), reverse=reverse)
        
        # Paginate
        paginated = results[offset:offset + limit]
        
        return {
            "codes": [self._format_code_summary(c) for c in paginated],
            "total": len(results),
            "limit": limit,
            "offset": offset,
            "hasMore": offset + limit < len(results),
        }
    
    def get_code(self, code: str) -> Optional[Dict[str, Any]]:
        """Get a single code by code string"""
        code_obj = self.code_index.get(code) or self.code_index.get(code.upper())
        if not code_obj:
            return None
        return self._format_code_detail(code_obj)
    
    def search_codes(
        self,
        query: str,
        limit: int = 50,
        code_type: Optional[str] = None
    ) -> Dict[str, Any]:
        """Search codes by query string"""
        if not query or len(query.strip()) < 2:
            return {"codes": [], "total": 0, "query": query}
        
        search_term = query.lower().strip()
        terms = search_term.split()
        
        # Score-based search
        results = []
        
        for item in self.search_index:
            # Filter by type if specified
            if code_type and item["type"] != code_type.upper():
                continue
            
            score = 0
            
            # Exact code match gets highest score
            if item["code"].lower() == search_term:
                score = 100
            elif search_term in item["code"].lower():
                score = 80
            
            # Term matching in description
            for term in terms:
                if term in item["search_text"]:
                    score += 10
            
            if score > 0:
                results.append({"code": item["code"], "score": score})
        
        # Sort by score descending
        results.sort(key=lambda x: x["score"], reverse=True)
        
        # Get full code objects for top results
        top_results = []
        for r in results[:limit]:
            code_obj = self.code_index.get(r["code"])
            if code_obj:
                top_results.append(self._format_code_summary(code_obj))
        
        return {
            "codes": top_results,
            "total": len(results),
            "query": query,
        }
    
    def _format_code_summary(self, code_obj: Dict[str, Any]) -> Dict[str, Any]:
        """Format code for summary listing"""
        try:
            code = Code.from_raw(code_obj)
            return code.to_summary()
        except Exception:
            # Fallback formatting
            return {
                "code": code_obj.get("code"),
                "description": code_obj.get("description"),
                "category": code_obj.get("labels", [None])[0] if code_obj.get("labels") else self._normalize_type(code_obj.get("type")),
                "type": self._normalize_type(code_obj.get("type")),
                "labels": code_obj.get("labels", []),
            }
    
    def _format_code_detail(self, code_obj: Dict[str, Any]) -> Dict[str, Any]:
        """Format code for detailed view with payments"""
        try:
            code = Code.from_raw(code_obj)
            return code.to_detail()
        except Exception:
            # Fallback formatting
            return {
                "code": code_obj.get("code"),
                "description": code_obj.get("description"),
                "category": code_obj.get("labels", [None])[0] if code_obj.get("labels") else self._normalize_type(code_obj.get("type")),
                "type": self._normalize_type(code_obj.get("type")),
                "labels": code_obj.get("labels", []),
                "payments": {"IPPS": 0, "HOPD": 0, "ASC": 0, "OBL": 0},
                "optional": {},
                "rawMetadata": code_obj.get("metadata"),
            }
    
    def get_stats(self) -> Dict[str, Any]:
        """Get statistics about loaded codes"""
        stats = {
            "totalCodes": len(self.codes),
            "isLoaded": self._is_loaded,
            "types": {},
            "loadMethod": "chunked" if self.manifest else "single-file",
        }
        
        if self.manifest:
            stats["chunks"] = {
                "count": self.manifest.get("chunkCount"),
                "targetSizeMB": self.manifest.get("targetChunkSizeMB"),
                "createdAt": self.manifest.get("createdAt"),
            }
        
        for type_name, codes in self.type_index.items():
            stats["types"][type_name] = len(codes)
        
        return stats
    
    def is_ready(self) -> bool:
        """Check if service is ready"""
        return self._is_loaded and self.load_error is None


# Singleton instance
code_service = CodeService()

